File: src/WagsMediaRepository.Domain/Models/VideoGame.cs
using System.Linq;

namespace WagsMediaRepository.Domain.Models;

public class VideoGame
{
    public int VideoGameId { get; set; }
    
    public string Title { get; set; } = string.Empty;
    
    public string Link { get; set; } = string.Empty;
    
    public DateTime? DateStarted { get; set; }
    
    public DateTime? DateCompleted { get; set; }
    
    public int Rating { get; set; }

    public string Thoughts { get; set; } = string.Empty;
    
    public string CoverImageUrl { get; set; } = string.Empty;
    
    public int? SortOrder { get; set; }

    public Constants.VideoGameStatus Status { get; set; } = Constants.VideoGameStatus.ToPlay;

    public Constants.VideoGameCompletionStatus CompletionStatus { get; set; } = Constants.VideoGameCompletionStatus.NotApplicable;

    public List<VideoGameGenre> Genres { get; set; } = [];
    
    public List<VideoGameSystem> Systems { get; set; } = [];

    public static VideoGame FromDto(VideoGameDto dto) => new()
    {
        VideoGameId = dto.VideoGameId,
        Title = dto.Title,
        Link = dto.Link,
        DateStarted = dto.DateStarted,
        DateCompleted = dto.DateCompleted,
        Rating = dto.Rating,
        Thoughts = dto.Thoughts,
        CoverImageUrl = dto.CoverImageUrl,
        SortOrder = dto.SortOrder,
        Status = (Constants.VideoGameStatus)dto.VideoGameStatusId,
        CompletionStatus = (Constants.VideoGameCompletionStatus)dto.VideoGameCompletionId,
        Genres = dto.VideoGameToVideoGameGenres
            .Select(vg => vg.VideoGameGenre)
            .Select(VideoGameGenre.FromDto)
            .ToList(),
        Systems = dto.VideoGameToVideoGameSystems
            .Select(vs => vs.VideoGameSystem)
            .Select(VideoGameSystem.FromDto)
            .ToList(),
    };
}

File: api/src/models/videoGame.ts
export type VideoGameSystem = {
	videoGameSystemId: number;
	name: string;
	colorCode: string;
	videoGameCount?: number;
}

export type VideoGameGenre = {
	videoGameGenreId: number;
	name: string;
	colorCode: string;
	videoGameCount?: number;
}

export type VideoGameStatus = {
	videoGameStatusId: number;
	name: string;
	colorCode: string;
}

export type VideoGameCompletion = {
	videoGameCompletionId: number;
	name: string;
	colorCode: string;
}

export type VideoGame = {
	videoGameId: number;
	videoGameStatusId: number;
	videoGameCompletionId: number;
	title: string;
	link: string;
	dateStarted?: Date | string | null;
	dateCompleted?: Date | string | null;
	rating: number;
	thoughts: string;
	coverImageUrl: string;
	sortOrder?: number;
	status?: VideoGameStatus;
	completion?: VideoGameCompletion;
	genres: VideoGameGenre[];
	systems: VideoGameSystem[];
}

export type VideoGameGenreLink = {
	videoGameId: number;
	genreId: number;
	genreName: string;
	genreColorCode: string;
}

export type VideoGameSystemLink = {
	videoGameId: number;
	systemId: number;
	systemName: string;
	systemColorCode: string;
}

export type VideoGameQueryReturn = {
	VideoGameId: number;
	VideoGameStatusId: number;
	VideoGameCompletionId: number;
	Title: string;
	Link: string;
	DateStarted?: Date;
	DateCompleted?: Date;
	Rating: number;
	Thoughts: string;
	CoverImageUrl: string;
	SortOrder?: number;
	VideoGameStatusName: string;
	VideoGameStatusColor: string;
	VideoGameCompletionName: string;
	VideoGameCompletionColor: string;
}

export type VideoGameGenreLinkQueryReturn = {
	VideoGameId: number;
	VideoGameGenreId: number;
	VideoGameGenreName: string;
	VideoGameGenreColor: string;
}

export type VideoGameSystemLinkQueryReturn = {
	VideoGameId: number;
	VideoGameSystemId: number;
	VideoGameSystemName: string;
	VideoGameSystemColor: string;
}

export type VideoGameGenreQueryReturn = {
	VideoGameGenreId: number;
	Name: string;
	ColorCode: string;
	VideoGameCount: number;
}

export type VideoGameSystemQueryReturn = {
	VideoGameSystemId: number;
	Name: string;
	ColorCode: string;
	VideoGameCount: number;
}
